lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp

```diff
@@ -1092,9 +1096,223 @@ class DecomposeAtenMvOp : public OpRewritePattern<AtenMvOp> {
 };
 } // namespace
 
+// Decompose aten.pixel_shuffle into: aten.permute and aten.reshape operations.
+//
+// If input is a tensor of shape (*leading_dims, C*r*r, H, W), where
+// leading_dims is of size N, then
+//    X = pixel_shuffle(input, upscale_factor)
+//
+// gets replaced with
+//    A = input.reshape(*leading_dims, C, r, r, H, W)
+//    B = A.permute(0, ..., N, N+3, N+1, N+4, N+2)
+//    X = B.reshape(*leading_dims, C, r*H, r*W)
+//
+// 'r' above is referred to as the 'upscale factor' or just 'factor' below.
+namespace {
+class DecomposeAtenPixelShuffleOp
+    : public OpRewritePattern<AtenPixelShuffleOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(AtenPixelShuffleOp op,
+                                PatternRewriter &rewriter) const override {
+
+
+    Location loc = op.getLoc();
+    Value inValue = op.getSelf();
+    auto inType = inValue.getType().cast<BaseTensorType>();
+    auto maybeSizes = inType.getOptionalSizes();
+    if (!maybeSizes) {
+      return rewriter.notifyMatchFailure(
+          op, "Expected input tensor to have known rank.");
+    }
+    auto inShape = maybeSizes.value();
+    auto inRank = inShape.size();
+
+    // TODO: Handle dynamic shapes.
+    //
+    // Currently the decomposition of inputs with dynamic shapes results in a
+    // failure when lowering to linalg. Specifically, the error:
+    //
+    // "Unhandled case of expand/collapse. "
+    //
+    // is hit in the lowering of aten.view to the linalg dialect (see
+    // ConvertAtenViewOp in TorchToLinalg/DataMovement.cpp where this error is
+    // thrown).
+    //
+    // The issue is that there is no way for the lowering from torch to linalg
+    // to know that the view ops (reshape ops) created in this decomposition can
+    // be mapped directly to tensor.expand_shape and tensor_collapse_shape. We
+    // DO know that this is the case: the first reshape is an expand_shape
+    // and the second one is a collapse_shape. But we only know this because of
+    // the semantics of pixel_shuffle, by the time the lowering sees the view
+    // ops, this special context is lost.
+    //
+    // There are a few possible options to fixing the current approach, as I see
+    // it:
+    //
+    // 1) Lower pixel_shuffle to linalg directly. This would mean a bit of
+    // duplication of the logic for lowering from aten.permute to linalg, and
+    // would not be the most 'gradual' lowering approach. But perhaps the most
+    // straightforward and understable.
+    //
+    // 2) Rather than use aten.reshape ops in the decomposition below, insert
+    // tensor.expand_shape and tensor.collapse_shape ops directly. This would
+    // require inserting unrealized_conversion_cast ops to convert the types of
+    // the operands from !torch.tensor to to the MLIR built-in tensor type,
+    // which isn't ideal (unrealized_conversion_cast should only be used in
+    // conversion passes AFAIK (?)).
```

Comment:
Torch-mlir cannot decompose to a tensor operation. One important part of torch-mlir is the ability to lower to different target IRs (or even mix-and-match), thus all torch decompositions must be strictly in terms of torch ops.

In regards to `builtin.unrealized_conversion_cast` though, there is a `torchc.to_builtin` and `torchc.from_builtin` (might have gotten the names wrong) that allow conversion between the builtin tensor type and the torch tensor type.

Reply:
Makes sense, so this isn't an option because we it'd break the lowering to tosa, for example. Good to know that about the to_builtin and from_builtin.  Out of curiosity, do you know why torch-mlir decided against using the builtin tensor type to represent a value semantics tensor?

Reply:
I think it was some attempt to reflect the more general programming model that pytorch offers, and since a non-value semantics tensor was going to be needed anyway, I'm guessing the idea was that swapping between builtin types and a torch type would make things more difficult. I wasn't around for those design decisions though :)